WcaOnRails/app/models/locale.rb

```diff
@@ -0,0 +1,124 @@
+      decorate_with_hashes(file_content, self, "")
+    end
+  end
+
+  def compare_to(base)
+    compare_node_resursive(base[base.locale], self[locale], [])
+  end
+
+  private
+
+  # Adapted from https://github.com/jonatanklosko/internationalize/blob/7090c90d4d8e4571025c3be4484b5f668cbb6501/client/app/services/translation-utils.service.js#L195-L221
+  def decorate_with_hashes(text, node, prefix)
+    node.each do |key, value|
+      if leaf?(value)
+        # We want leaves to have at least "_translated", and maybe later "_hash"
+        node[key] = { "_translated" => value }
+      else
+        text = decorate_with_hashes(text, value, "#{prefix}#{SOME_CHARS}#{KEY_MATCHER % key}")
+      end
+      regexp = Regexp.new "(#{prefix}#{SOME_CHARS})#{COMMENT_LINE_GROUP}\\s*#{KEY_MATCHER % key}"
+      text = text.sub(regexp) do
+        # $1 is everything before the comment and the key
+        before = $1
+        # $2 is the hash(es) matched, or empty
+        hashes = $2
+        unless hashes.empty?
+          hash_lines = hashes.split('#').map(&:strip!)
+          if hash_lines
+            hash = hash_lines.select(&method(:line_filter)).map(&method(:line_cleaner))
+            if hash.size >= 1
+              node[key]["_hash"] = hash.first
+            end
+          end
+        end
+        # We return the beginning without the key, so that the current hash + key are removed from the text,
+        # but the parents and the value stay in.
+        before
+      end
+    end
+    text
+  end
+
+  def compare_node_resursive(base, translation, context)
+    missing_keys = []
+    outdated_keys = []
+    base.each do |key, value|
+      unless translation.key?(key)
+        if leaf?(value)
+          missing_keys << fully_qualified_name(context, key)
+        else
+          context.push(key)
+          missing_keys += get_all_recursive(value, context)
+          context.pop
+        end
+        next
+      end
+      if leaf?(value)
+        unless Digest::SHA1.hexdigest(value)[0..6] == translation[key]["_hash"]
+          outdated_keys << fully_qualified_name(context, key)
+        end
+      else
+        context.push(key)
+        missing, outdated = compare_node_resursive(value, translation[key], context)
+        context.pop
+        missing_keys += missing
+        outdated_keys += outdated
+      end
+    end
+    [missing_keys, outdated_keys]
+  end
+
+  def get_all_recursive(node, context)
+    leaves = []
+    node.each do |key, value|
+      if leaf?(value)
+        leaves << fully_qualified_name(context, key)
+      else
+        context.push(key)
+        leaves += get_all_recursive(value, context)
+        context.pop
+      end
+    end
+    leaves
+  end
+
+  def fully_qualified_name(context, key)
+    # Some keys are actually numbers
+    key = key.to_s
+    # Some keys are too long and would overflow from the list-item, shrink it to an arbitrary length
+    if key.length > 20
+      key = key[0...19] + "..."
+    end
```

Comment:
Look at [this method](http://api.rubyonrails.org/classes/ActionView/Helpers/TextHelper.html#method-i-truncate) =)